### PR TITLE
spack compiler find: write into packages.yaml

### PR DESCRIPTION
**Pull request by alalazo**
This commit makes:
```console
$ spack compiler find 
```
write in `packages.yaml`, instead of `compilers.yaml`. It also removes dead code after this switch.

### CODE REVIEW

**Comment by alalazo** [comment]
It seems these compilers are detected on Windows:
```yaml
packages:
  llvm:
    externals:
    - spec: llvm@16.0.6+clang+lld~lldb
      prefix: C:\Program Files\LLVM
      extra_attributes:
        compilers:
          c: C:\Program Files\LLVM\bin\clang.exe
          cxx: C:\Program Files\LLVM\bin\clang++.exe
          ld: C:\Program Files\LLVM\bin\ld.lld.exe
  gcc:
    externals:
    - spec: gcc@8.3.0 languages='c,c++,fortran'
      prefix: C:\Strawberry\c
      extra_attributes:
        compilers:
          c: C:\Strawberry\c\bin\gcc.exe
          cxx: C:\Strawberry\c\bin\g++.exe
          fortran: C:\Strawberry\c\bin\gfortran.exe
    - spec: gcc@12.2.0 languages='c,c++,fortran'
      prefix: C:\mingw64
      extra_attributes:
        compilers:
          c: C:\mingw64\bin\gcc.exe
          cxx: C:\mingw64\bin\g++.exe
          fortran: C:\mingw64\bin\gfortran.exe
```

**Comment by alalazo** [comment]
Apparently one needs to source:
```
. "C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\Tools\Launch-VsDevShell.ps1"
```
to get compilers in PATH on `windows-latest`.

**Comment by haampie** [comment]
Why is `ld: C:\Program Files\LLVM\bin\ld.lld.exe` added?

**Comment by alalazo** [comment]
> Why is ld: C:\Program Files\LLVM\bin\ld.lld.exe added?

Not part of this PR. It's the external search logic in LLVM that finds it.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml
Comment (haampie) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml:1]: undo unrelated changes?

File: share/spack/gitlab/cloud_pipelines/stacks/aws-isc/spack.yaml
Comment (haampie) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-isc/spack.yaml:1]: same

File: share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml
Comment (alalazo) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml:1]: Having a float as version was causing failures in pipelines. I can extract the changes but this was needed to pass tests.

File: share/spack/gitlab/cloud_pipelines/stacks/aws-isc/spack.yaml [hunk 1]
```diff
@@ -39,13 +39,17 @@ spack:
     openmpi:
       variants: fabrics=ofi +legacylaunchers
     openturns:
-      version: [1.18]
+      version:
+      - '1.18'
```
Comment (alalazo) [inline share/spack/gitlab/cloud_pipelines/stacks/aws-isc/spack.yaml]: This is the change that was needed

File: share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml
Comment (haampie) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml:1]: Hm ok, I was thinking it makes it harder to revert if necessary.

Comment (haampie) [reply share/spack/gitlab/cloud_pipelines/stacks/aws-isc-aarch64/spack.yaml:1]: I've merged it separately

**Comment by haampie** [comment]
1. This adds odd, non-standard paths for ubuntu system compilers:

   ```yaml
   - spec: gcc@10.5.0 languages='c,c++,fortran'
     prefix: /usr
     extra_attributes:
       compilers:
         c: /usr/bin/x86_64-linux-gnu-gcc-10
         cxx: /usr/bin/x86_64-linux-gnu-g++-10
         fortran: /usr/bin/x86_64-linux-gnu-gfortran-10
   ```
   
   used to be:
   
   ```yaml
   - compiler:
       spec: gcc@=10.5.0
       paths:
         cc: /usr/bin/gcc-10
         cxx: /usr/bin/g++-10
         f77: /usr/bin/gfortran-10
         fc: /usr/bin/gfortran-10
       flags: {}
       operating_system: ubuntu23.10
       target: x86_64
       modules: []
       environment: {}
       extra_rpaths: []
   ```
   
   Also inconsistencies for clang:
   
   ```yaml
   - spec: llvm@16.0.6+clang+lld~lldb
     prefix: /usr
     extra_attributes:
       compilers:
         c: /usr/bin/clang
         cxx: /usr/bin/clang++-16
         ld: /usr/bin/ld.lld
   ```

2. There is no schema validation of `extra_attributes:compilers`, which is very annoying, especially since there is no suggestion how/where to add `flags` / `environment` / `extra_rpaths`.
   
3. Not sure about listing the executable path to lld (and lldb), it's not a compiler, and it's confusing since that linker is not used by default, which the config suggests in a way.

4. It's significantly slower (about 5x) than the original `spack compiler find`

To me it feels like a regression in its current state, even without trying `--mixed-toolchain` which I guess becomes unsupported...

**Comment by alalazo** [comment]
> even without trying --mixed-toolchain which I guess becomes unsupported...

That should work, but will add e.g. `gcc` executables to `llvm` specs. The flag will disappear when compiler becomes nodes.

> This adds odd, non-standard paths for ubuntu system compilers:

This can be changed in a separate PR to be merged before this one. @becker33 Any opinion on this topic? I think it has been like this since ever, and was not changed in #43464

> There is no schema validation of extra_attributes:compilers, which is very annoying, especially since there is no suggestion how/where to add flags / environment / extra_rpaths.

We should rework the docs. Forgot to mention that point above. We can possibly add some schema validation based on regex matching.

> Not sure about listing the executable path to lld (and lldb), it's not a compiler, and it's confusing since that linker is not used by default, which the config suggests in a way.

That can be moved to another section of `extra_attributes`. Can be done in a separate PR too.

File: lib/spack/spack/repo.py [hunk 2]
```diff
@@ -852,7 +852,8 @@ def get(self, spec):
 
     def get_pkg_class(self, pkg_name):
         """Find a class for the spec's package and return the class object."""
-        return self.repo_for_pkg(pkg_name).get_pkg_class(pkg_name)
+        with _switch_global_repo(self):
+            return self.repo_for_pkg(pkg_name).get_pkg_class(pkg_name)
```
Comment (alalazo) [inline lib/spack/spack/repo.py]: This is another change that needs to be carefully vetted, and that is mainly needed for unit tests. 

`spack external find` uses processes, and those have a different spawning policy depending on the platform. Before, `spack compiler find` was using threads so we didn't have this issue. Maybe we should find a way to enforce these modifications only in tests. :thinking:

Comment (alalazo) [reply lib/spack/spack/repo.py]: This was solved in a better way in https://github.com/spack/spack/pull/45053, so removed from this PR

**Comment by alalazo** [comment]
The failure in https://github.com/spack/spack/actions/runs/9986088746/job/27598142886?pr=44419 seems to be due to:
1. That test triggering an automatic compiler find, which is written to `packages.yaml`
2. `gcc languages=c,c++,fortran,d` being detected as a compiler
3. The spec above not being valid after we switched the `builtin` repo with `builtin.mock`
4. Spack failing to concretize when any external spec is invalid

I still have to verify that `macOS-13` has indeed a `d` compiler, and the other instances do not. Anyhow, this is an unrelated bug we also have on develop, see  #45321

**Review by haampie — CHANGES_REQUESTED** [review]
This breaks user workflows.

You can no longer rely on `compilers:: [...]` to drop compilers from user config.

Workarounds for that are really bad as you have to exhaustively list every compiler package in packages config additionally:

```yaml
spack:
  compilers:: []
  packages:
    gcc:
       externals:: []
    llvm: # initially I put clang here... which certainly the majority of the users will do too
       externals:: []
```

(On top  of that you can't prevent Spack from registering clang as compiler if you just wanted to expose libLLVM.so, but that's not this PR's fault)

---

Also there is a bug where Spack still modifies the top-level config scope even if no new compilers are added, like this:

```yaml
$ cat spack.yaml 
spack:
  'compilers:': []
  packages:  # all this redundant stuff is added
    gcc:
      externals: []
    llvm:
      externals: []
```

**Comment by alalazo** [comment]
> This breaks user workflows. You can no longer rely on compilers:: [...] to drop compilers from user config.

We need to figure out a mitigation for this, but this behavior is already in v0.22 (it just won't be triggered that frequently)

**Comment by alalazo** [comment]
> Workarounds for that are really bad as you have to exhaustively list every compiler package in packages config additionally:

I guess we can also use `require:`, but that is also not a zero cost workaround. 

One thing I can propose is to reconsider the approach that we ditched in March / April and that was prototyped in back in #17525. Basically:
1. Have a first PR to `develop` that only unifies the search mechanisms, 
2. Leave the switch to `packages.yaml` for #45189, when ready. 
 
In retrospect, if we now go with that it would have been good to NOT allow compilers from `packages.yaml`, but I guess that is a ship that sailed already. This might be a good topic for this Thursday meeting. @becker33 @tgamblin FYI

**Comment by psakievich** [comment]
I have some questions.

How will mixed toolchains be accounted for?

Say I have a compiler that only supplies `languages=[c,cxx]` like `apple-clang`.  Is the expectation to manually add a `fortran` entry when needed and then still use `foo%apple-clang` if `foo` `depends_on("fortran")` or would I do something like`foo%apple-clang.c%apple-clang.cxx%gcc.fortran`?  Also is double `%`  syntactically correct here?  I'm already forgetting some of these nuances.

Seems like the former will work for many cases and meets some of my requests for a concise way to represent a tool chain. However, it feels really hacky and also seems super limited.  The latter is really verbose, but also very clear.  Anything in the middle seems scary like `foo%apple-clang%gcc`.  I suppose `foo%apple-clang%gcc.fortran` would be a little better if the concretizer were smart enough to know and enforce that `apple-clang.[c,cxx]` need to take precedent over `gcc.[c,cxx]`.  

Anyways what's the correct syntax here?

**Comment by alalazo** [comment]
@psakievich Syntax changes are discussed in https://github.com/spack/spack/issues/44379 (I see a :+1: from you there :wink: ) There's nothing like `%gcc.cxx` or similar syntax, so not sure where this is coming from. Using a dot is ambiguous: `gcc.fortran` right now parses as the `fortran` package from the `gcc` repository.

The `%` gets translated to "direct build dependency".

To you question on mixed toolchain, the example we have in the issue reads:
```console
foo %apple-clang %[virtuals=fortran] gcc
```

**Comment by psakievich** [comment]
> @psakievich Syntax changes are discussed in #44379 (I see a 👍 from you there 😉 ) 

Sorry. Too many sources of input to keep track of. I'll review and move my comments there.

**Comment by alalazo** [comment]
Superseded by #45784

**Pull request closed without merging**